### DISCUSSION TITLE
Easy-to-use, No-Dependencies Grammar Compiler for better response.

### DISCUSSION THREAD

**Discussion by nova-land** [ggml-org/llama.cpp#3729] [Show and tell]
I'd like to introduce a user-friendly Python-based GBNF Compiler, which you can easily access at: 
  - https://pypi.org/project/gbnf-compiler/
  - https://github.com/nova-land/gbnf-compiler/

The GBNF Compiler work as follow:

> Create a Template -> Provide Grammar String to LLM -> Parse the LLM response directly

While JSON-based grammar tools are excellent, they may not always be the ideal format for Large Language Models to read in. With this in mind, I developed the GBNF Compiler, designed to be as straightforward as using handlebars.js with double quotes. This flexibility allows you to adapt it to any grammar format you require.

With the GBNF Compiler, you can effortlessly create 'fill in the blank' grammars, perfect for LLMs to perform reasoning and explanation tasks. This approach enables us to achieve higher-quality and more easily parsable results."

------

Here with an example:

```python
import requests
from gbnf_compiler import GBNFCompiler, MultipleChoice, SingleSentence

# Define your Prompt
prompt = "What tool will you use to calculate 2^5 ?"

# Define the LLM Response Template
# Each {{}} is a variable with a rule
template = "I choose {{tool}} because {{reason}}"

tools = MultipleChoice('tool', ['calculator', 'web-search', 'web-browse'])
c = GBNFCompiler(template, { 'tool': tools, 'reason': SingleSentence() })
print(c.grammar())

# Try a dummy result
text = "I choose calculator because it is the most efficient and accurate way to calculate 2^5."
result = c.parse(text)
print(result)

# Example: Send it out to local llama.cpp
def template(role: str, prompt: str):
    return """[INST] <<SYS>>
{role}
<</SYS>>
{prompt}
[/INST]""".format(role=role, prompt=prompt)

data_json = {
    "prompt": template("", prompt), "temperature": 0.0,
    "n_predict": 512, "top_p": 0.2, "top_k": 10,
    "stream": False, "grammar": c.grammar() }

resp = requests.post(
    url="http://127.0.0.1:9999/completion",
    headers={"Content-Type": "application/json"},
    json=data_json,
)
result = resp.json()["content"]
print(result)

# Directly parse the result
# Try a dummy result
text = "I choose calculator because it is the most efficient and accurate way to calculate 2^5."
result = c.parse(text)
print(result)
"""
Result: 
{'tool': 'calculator', 'reason': 'it is the most efficient and accurate way to calculate 2^5.'}
"""

```

You can check more examples in [here](https://github.com/nova-land/gbnf-compiler/blob/main/tests/test_compiler.py) with JSON format grammar.

**Comment by arthurwolf** [comment]
this is absolutely brilliant

**Comment by nabil6391-seekasia** [comment]
Thats great, would it be possible to output a multi label classifcation, like I choose calculator or web-search

**Reply by nova-land** [reply]
Yes, it can create such output just like the ItemList Rule. However, I can't find any GBNF documentation that supports regex-like item limitation similar to `item{2}`, so I doubt it would be feasible to limit it into exactly <= 2 choices.

**Comment by bozo32** [comment]
hi
does the below qualify as a detailed use case?
I copied over and butchered, I'm sure, a script produced by [paschembri](https://github.com/paschembri)

Peter.tamas at wur.nl is the best way to find me

before I get there
there are differently challenging methods I can think of
metaphor analysis (identification of plausible and choice of probable connotations to associate with a word whose literal meaning is different than active). The hard part here might be in consistently identifying figurative language, identifying a plausible set of connotations and then, from those, appropriately referencing context to determine the probable (e.g. When a leader of the BJP calls a congress candidate 'regal' in the context of an election campaign is means something quite different than when a brit does it in the context of a wedding)

Frame analysis within a news article. (which would be kinda-sorta huge for analysis of political speeches etc.)
This would be identification of victims of a problem, those whose behaviour produces the problem, those who are responsible for the behaviour that harms, those who are responsible to fix a problem. 
For this one I have a PhD thesis that might be fun to replicate using LLMs....the work is here (https://research.wur.nl/en/publications/chickens-inc-was-uk-newspapers-framing-of-the-chicken-meat-produc) and the dataset is here (https://ssh.datastations.nl/dataset.xhtml?persistentId=doi:10.17026/dans-xuq-ve6a). 

Argument assessment. once an argument analysis is completed, it is possible to assess its merits. This would be a follow on to argument analysis (what the hell are they doing) then (how well are they doing it)

I have a bunch of MSc students running analyses of the responses in this journal issue now
https://journals.sagepub.com/toc/jcec/42/4
it would be fun to do a comparison between human and machine attempts.
They will be submitting their work in mid December

Now for my workup of a stripped out argument analysis. I tried to do the recursive structure...probably badly.

I will be using this stuff in MSc intro to qual analysis classes in January (whatever I can get working)

this sample text is drawn from
https://www.government.nl/documents/speeches/2020/03/16/television-address-by-prime-minister-mark-rutte-of-the-netherlands

# -*- coding: utf-8 -*-
from typing import List
from dataclasses import dataclass
from llm_core.assistants import LLaMACPPAssistant


sample_paper = """
Good evening,

Coronavirus has our country in its grip.

Our country and the entire world.

Together we face an enormous task.

In recent weeks many of us have felt like we’re on a rollercoaster that seems to be going faster and faster.

Asking ourselves: is this really happening?

After all, the measures taken here and elsewhere are unprecedented in peacetime.

I’d like to begin by expressing my sympathies to the families of those people who have died from the virus.

And I’d like to wish all the very best to everyone who is in hospital or is recuperating at home.

And let me say to those members of the public who are older, or have underlying health issues: I realise that you are very concerned.

That’s why I want to assure you that our top priority is to reduce the risk to you as much as we possibly can.

Given all the news both at home and abroad, and all the events unfolding at record speed, it’s entirely understandable that people everywhere are concerned.

We all have questions.

What can I do to protect myself and those close to me?

What about school and work?

Can your child’s birthday party go ahead? Your family weekend? Your wedding?

How long is this going to go on for?

And why are different countries taking different measures?

The reality is that coronavirus is here in our midst, and for the time being it is here to stay.

In today’s world, news and information travel faster than the speed of light, and people are quick to share their opinions.

I understand that.

But in order to answer the many questions people have, we need the knowledge and experience of experts.

Let us rely on that knowledge and experience.

And on experts such as Jaap van Dissel and his colleagues both at the National Institute for Public Health and the Environment and elsewhere.

On the virologists, intensive care doctors and other specialists.

Right from the very beginning, their advice has been the guiding force behind the measures we’ve taken in the Netherlands so far.

And it’s important that we continue to be guided by scientific knowledge and reliable facts.

That is the only sensible way in which we can keep taking the necessary steps.

Steps which inevitably lie ahead.

Because, my message to you this evening is not an easy one to hear.

The reality is that coronavirus is here in our midst, and for the time being it is here to stay.

There is no quick or easy solution to this intensely difficult situation.

The reality is also that in the coming period a large proportion of the Dutch population will become infected with this virus.

That’s what the experts are telling us now.

They are also telling us that – as we wait for a vaccine or treatment to be developed – we can delay the spread of the virus and at the same time build up population immunity in a controlled manner.

Let me explain what that entails.

Anyone who has had the virus is usually immune to it afterwards.

Just like with the measles back in the day.

The bigger the group that acquires immunity, the smaller the chance that the virus can make the leap to vulnerable older people or people with underlying health issues.

The aim of population immunity is to build, as it were, a protective wall around this group.

That’s the principle.

But it’s important to realise that it can take months, or even longer to build up population immunity, and in the meantime we have to protect high-risk groups as much as possible.

If we consider the big picture, there are three possible approaches.

The first is: control the virus as much as possible.

That should lead to a controlled spread among the groups least at risk.

That is the approach we have chosen.

Maximum control means taking measures aimed at reducing the peak in infections and staggering those infections over a longer period.

Maximum control means taking measures aimed at reducing the peak in infections and staggering those infections over a longer period.

By taking this approach, one in which most people will experience only minor symptoms, we can both build immunity and ensure that our healthcare system is able to cope.

So that our nursing homes, home care services, hospitals and, above all, our intensive care units don’t become overwhelmed.

And so that there’s always enough capacity to help the people who need it most.

The second option is that we simply allow the virus to run its course.

If we did that, our healthcare system would be completely swamped when infections peaked, so there wouldn’t be enough capacity to help vulnerable older people and other high-risk patients.

And that, of course, is a scenario we must prevent at all cost.

The third option is that we keep working endlessly to contain the virus.

That would mean shutting down the country completely.

Such a rigorous approach may seem like an attractive option, but experts say that this would not be a matter of days or weeks.

In this scenario, we would essentially have to shut the country down for a year or even longer, with all the consequences that would entail.

And even if that were possible in practice – making people stay in their homes unless they have permission to go outside, for such a lengthy period – the virus could simply rear its head again once the measures were lifted.

The Netherlands is an open country.

Until a vaccine is available, coronavirus will continue to sweep the world, and it won’t spare the Netherlands.

All the advice we’ve received, and all the measures we’ve announced so far, have been aimed at the first approach: controlling the spread as much as possible.

From relatively simple guidelines, such as not shaking hands, washing our hands more often and keeping a distance of one-and-a-half metres, to far-reaching measures such as banning large events and closing bars, clubs and restaurants.

And of course, we are constantly reviewing the situation.  

How long these measures need to be in place and whether more measures are necessary will depend on how the virus behaves in the coming weeks and months.

And on new scientific findings, because the research continues apace.

It’s possible that some measures can be relaxed, but that we may also have to take new steps to prevent the virus proceeding unchecked.

In the months ahead it will be a matter of adjusting our approach as necessary.

Of striking the right balance between taking measures that are needed and allowing normal life to continue as much as possible.

If we can control the spread of the virus in this way, the public health consequences will be easier to manage in the long run.

At the same time, we cannot and will not ignore the economic impact of this crisis.

Many people are concerned about their jobs.

Because, for many companies, large and small, this is an extremely difficult time.

Many businesspeople suddenly have their backs to the wall.

The lady with the coffee bar on the corner, the flower grower, the independent haulier, the self-employed person.

But even national icons like KLM are feeling the strain.

My message to all business owners in the Netherlands and to all their employees is this: the government will do what is necessary to support you.

We will do everything in our power to ensure that companies do not go under because of this crisis, and that people do not lose their jobs.

Whatever happens, this will be a difficult time.

But we will not let you down.

Lastly, I would like to thank everyone in the Netherlands for the way in which they have heeded the instructions and measures announced so far, and for all the heartwarming examples we’ve seen of people offering each other their help and solidarity.

It’s encouraging to see that when the need arises we are ready to help each other.

Please keep doing that.

Stay alert and follow the instructions, even if you are fit and healthy, for the sake of those who are more vulnerable.

That is extremely important.

Keep using your common sense and listen to what the experts say.

Keep helping each other where possible.

This is a time in which we must find common ground, overcoming our differences and divisions.

A time in which we must put the common interest above our own self-interest.

A time in which we must put our trust in those who are working day and night in difficult conditions to help others and get the virus under control.

We must give them space to work. Cleaners, nurses and doctors in hospitals and elderly care homes, GPs and municipal health workers, police officers, ambulance crews and other emergency workers.

To them, and to all the other people who have stayed at their posts in schools, childcare centres, public transport, supermarkets and elsewhere, I want to say: you are doing a fantastic job – thank you very, very much.

I would like to close with an appeal to you all: despite all the uncertainty, one thing is perfectly clear: the challenge we face is enormous, and all 17 million of us will have to work together to overcome it.

Together we will get through this difficult period.

Take care of each other.

I’m counting on you.

Thank you.
"""


@dataclass
class ToulminArgumentAnalysis:
    system_prompt = "You are a argumentation analysis expert"
    prompt = """
    Developed by philosopher Stephen E. Toulmin, the Toulmin
    method is a style of argumentation that breaks arguments down into six
    component parts:

    - claim
    - grounds
    - warrant
    - qualifier
    - rebuttal
    - backing.

    In Toulmin’s method, every argument begins with three fundamental parts that will be the focus of this analysis:
    - the claim
    - the grounds
    - the warrant

    A claim is the assertion that authors would like to prove to their audience.
    The grounds of an argument are the evidence, the facts or the data that is used to support the claim. For example, in the sentence 'it will be sunny, so we should go to the beach' the assertion 'it will be sunny' are grounds'. These are normally stated by are sometimes implicit.
    The warrant, which is either implied or stated explicitly, is the rule or the assumption that makes the grounds relevant to the claim. For example, the ground 'it will be sunny' supports the claim 'we should go to the beach' if we know both that some days are cloudy and that being sunny makes it enjoyable to be at the beach. Quite often, as it is in this case, warrants are not stated. They require the reader to draw from their own understanding to infer the rule that makes the grounds relevant to the claim. 
    An argument is a claim + warrants + grounds
    In a simple argument, such as the example given, there may be only one layer. In a complex argument there may be many layers. For example, in the statement 'I checked the weather forecast and it will be sunny, so we should go to the beach' 'I checked the weather' are grounds that support the assertion 'it will be sunny'. In this case the unstated assumptions are that weather forecasts are accurate and that the speaker got the weather forecast for the beach on the right day. This complex argument should be represented in the following way:
    <argument1>
        <claim> we should go to the beach
        <ground1> it will be sunny
        <warrant1> [sunny days make the beach nice]
        <warrant2> [some days are not sunny]
    <argument2>
        <supporting> argument1-ground1
        <claim> it will be sunny
        <ground1> I checked the weather forecast
        <ground2> [the weather forecast is for the right beach on the right day]
        <warrant1> [weather forecasts are accurate]
    Where 
    a. <argument(1,2...)> identifies the argument number 
    b. [square brackets] indicate content that has been inferred.
    c. <claim> indicates the claim of the argument
    d. <ground(1,2...)> indicates a ground or data directly relevant to <claim>
    e. <warrant(1,2...)> indicates a warrant directly relevant to <claim>
    f. <supporting> indicates which component of which argument is directly supported


    NOTE: in this structure <argument2><claim>=<argument1><ground1>

    Argument analysis has the following steps:
    1. Pick the 'final claim' that you will analyse. This is the claim that is the main point of the whole text submitted for analysis. This is done by reading the whole article several times and deciding which claim is the most important. This claim, logically, is the conclusion that the entire text supports. This claim becomes <argument1><claim>
    2. Identify and record grounds <ground(1,2...)> and warrants <warrant(1,2...)> directly relevant to this claim
    3. If data or warrants are missing, infer their necessary content and record that in square brackets
    3. For each data and warrant that are present in the argument being examined copy it to <argument+1><claim> and repeat steps 2 and 3
    4. Work logically backward through the article recursively from the main claim mapping out the whole structure that supports <argument1>

    Here is a text we'd like to analyze:
    ```
    {content}
    ```

    > Perform a Toulmin analysis on the previous text.
    """
    argument: str
    claim: str
    ground: str
    warrant: str

    @classmethod
    def analyze(cls, content):
        with LLaMACPPAssistant(cls, model="mistral") as assistant:
            analysis = assistant.process(content=content)
            return analysis

if __name__ == "__main__":
    analysis = ToulminArgumentAnalysis.analyze(sample_paper)
    print(f"Claim: {analysis.claim}")
    print(f"Data: {analysis.data}")
    print(f"warrant: {analysis.warrant}")
